### PR TITLE
[Driver][SYCL] Improve dependency file behaviors for -fintelfpga

### PR DESCRIPTION
**Pull request by mdtoguchi**
When compiling AOT for FPGA dependency files are generated that are used
by the aoc compilation.  Single step compilation is seamless as the dependency
file is generated then immediately used.  When compiling to object, keeping
track of the dependency file is not so intuitive.

To alleviate problems of not being able to find the dependency file, the
original dependency file is bundled up with the destination fat object and
when used is unbundled and passed to the aoc compilation.

Signed-off-by: Michael D Toguchi <michael.d.toguchi@intel.com>

### CODE REVIEW

**Comment by mdtoguchi** [comment]
@bader, are clang-format failures required to be fixed?  This test must be new, as I didn't hit it with my previous changes and the bits are the same.

**Comment by bader** [comment]
> @bader, are clang-format failures required to be fixed? This test must be new, as I didn't hit it with my previous changes and the bits are the same.

I think so. I've recently added this check to align with LLVM community, which tries to keep consistent coding style across the code base by using similar pre-commit check.
You can fix this failure by applying the patch attached to the job artifacts.
https://github.com/intel/llvm/suites/481642712/artifacts/2232617

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 1]
```diff
@@ -6758,10 +6759,16 @@ const char *Clang::getDependencyFileName(const ArgList &Args,
...
-    SmallString<128> OutputFilename(OutputOpt->getValue());
+  auto makeDepName = [&](Arg *OutOpt) {
+    SmallString<128> OutputFilename(OutOpt->getValue());
     llvm::sys::path::replace_extension(OutputFilename, llvm::Twine('d'));
     return Args.MakeArgString(OutputFilename);
+  };
+  if (Arg *OutputOpt = Args.getLastArg(options::OPT_o)) {
+    return makeDepName(OutputOpt);
+  }
+  if (Arg *OutputOpt = Args.getLastArg(options::OPT__SLASH_Fo)) {
```
Comment (AGindinson) [inline clang/lib/Driver/ToolChains/Clang.cpp]: AFAIU, the variadic nature of the function can be re-used:
```suggestion
  if (Arg *OutputOpt = Args.getLastArg(options::OPT_o,
                                                               options::OPT__SLASH_Fo)) {
```

Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/Clang.cpp]: nice catch.  I have updated.

File: llvm/lib/Support/Triple.cpp [hunk 2]
```diff
@@ -399,69 +404,70 @@ static Triple::ArchType parseARMArch(StringRef ArchName) {
...
-    .Default(Triple::UnknownArch);
+                .Cases("i386", "i486", "i586", "i686", Triple::x86)
+                // FIXME: Do we need to support these?
+                .Cases("i786", "i886", "i986", Triple::x86)
+                .Cases("amd64", "x86_64", "x86_64h", Triple::x86_64)
+                .Cases("powerpc", "powerpcspe", "ppc", "ppc32", Triple::ppc)
+                .Cases("powerpc64", "ppu", "ppc64", Triple::ppc64)
+                .Cases("powerpc64le", "ppc64le", Triple::ppc64le)
+                .Case("xscale", Triple::arm)
+                .Case("xscaleeb", Triple::armeb)
```
Comment (AGindinson) [inline llvm/lib/Support/Triple.cpp]: @bader, is a large formatting change like this fine under the circumstances? No objections from me, just to confirm this is in accordance with your earlier request.

File: clang/test/Driver/sycl-offload-intelfpga.cpp [hunk 3]
```diff
@@ -155,12 +155,65 @@
...
+// CHK-FPGA-DEP-FILES2: clang{{.*}} "-dependency-file" "[[INPUT2:.+\.d]]"
+// CHK-FPGA-DEP-FILES2-LIN: clang-offload-bundler{{.*}} "-type=o" "-targets=sycl-spir64_fpga-unknown-unknown-sycldevice,host-x86_64-unknown-linux-gnu,sycl-fpga_dep" {{.*}} "-inputs={{.*}}.bc,{{.*}}.o,[[INPUT2]]"
+// CHK-FPGA-DEP-FILES2-WIN: clang-offload-bundler{{.*}} "-type=o" "-targets=sycl-spir64_fpga-unknown-unknown-sycldevice-coff,host-x86_64-pc-windows-msvc,sycl-fpga_dep" {{.*}} "-inputs={{.*}}.bc,{{.*}}.obj,[[INPUT2]]"
+
+/// -fintelfpga dependency file test to object with output designator
+// RUN: touch %t-1.cpp
+// RUN: %clangxx -### -target x86_64-unknown-linux-gnu -fsycl -fintelfpga %t-1.cpp -c -o dummy.o 2>&1 \
+  // RUN:  | FileCheck -check-prefixes=CHK-FPGA-DEP-FILES3,CHK-FPGA-DEP-FILES3-LIN %s
+// RUN: %clang_cl -### -fsycl -fintelfpga %t-1.cpp -c -Fodummy.obj 2>&1 \
+  // RUN:  | FileCheck -check-prefixes=CHK-FPGA-DEP-FILES3,CHK-FPGA-DEP-FILES3-WIN %s
```
Comment (AGindinson) [inline clang/test/Driver/sycl-offload-intelfpga.cpp]: Less-than-minor:
```suggestion
// RUN: %clangxx -### -target x86_64-unknown-linux-gnu -fsycl -fintelfpga %t-1.cpp -c -o dummy.o 2>&1 \
// RUN:  | FileCheck -check-prefixes=CHK-FPGA-DEP-FILES3,CHK-FPGA-DEP-FILES3-LIN %s
// RUN: %clang_cl -### -fsycl -fintelfpga %t-1.cpp -c -Fodummy.obj 2>&1 \
// RUN:  | FileCheck -check-prefixes=CHK-FPGA-DEP-FILES3,CHK-FPGA-DEP-FILES3-WIN %s
```

Comment (mdtoguchi) [reply clang/test/Driver/sycl-offload-intelfpga.cpp]: fixed.

File: llvm/lib/Support/Triple.cpp (hunk 2 above)
Comment (bader) [reply llvm/lib/Support/Triple.cpp]: LLVM community is currently working on enabling pre-commit checks and one the checks is that patches are processed with clang-format.
I enabled the same check on our side to make upstreaming and further merging easier.
I suggest starting to upstream driver patches ASAP to avoid any concerns about formatting changes, conflicts, etc.

File: llvm/include/llvm/ADT/Triple.h [hunk 4]
```diff
@@ -45,59 +45,60 @@ class Triple {
...
+    amdil64,    // AMDIL with 64-bit pointers
+    hsail,      // AMD HSAIL
+    hsail64,    // AMD HSAIL with 64-bit pointers
+    spir,       // SPIR: standard portable IR for OpenCL 32-bit version
+    spir64,     // SPIR: standard portable IR for OpenCL 64-bit version
+    kalimba,    // Kalimba: generic kalimba
+    shave,      // SHAVE: Movidius vector VLIW processors
+    lanai,      // Lanai: Lanai 32-bit
+    wasm32,     // WebAssembly with 32-bit pointers
+    wasm64,     // WebAssembly with 64-bit pointers
```
Comment (AlexeySachkov) [inline llvm/include/llvm/ADT/Triple.h]: @mdtoguchi, why this change was done?

Right now comments look inconsistent and this might cause problems during pulldowns from llvm/llvm-project

Comment (bader) [reply llvm/include/llvm/ADT/Triple.h]: Please, read https://github.com/intel/llvm/pull/1186#discussion_r388260545.

Comment (AlexeySachkov) [reply llvm/include/llvm/ADT/Triple.h]: Ok, thanks!
